contracts/asset-proxy/contracts/src/bridges/DydxBridge.sol

```diff
@@ -0,0 +1,236 @@
+
+            // Scale amount, if conversion rate is set.
+            uint256 scaledAmount;
+            if (bridgeAction.conversionRateDenominator > 0) {
+                scaledAmount = amount
+                    .safeMul(bridgeAction.conversionRateNumerator)
+                    .safeDiv(bridgeAction.conversionRateDenominator);
+            } else {
+                scaledAmount = amount;
+            }
+
+            // Construct dydx action.
+            if (bridgeAction.actionType == BridgeActionType.Deposit) {
+                // Deposit tokens from the account owner into their dydx account.
+                actions[i] = _createDepositAction(
+                    depositFrom,
+                    scaledAmount,
+                    bridgeAction
+                );
+            } else if (bridgeAction.actionType == BridgeActionType.Withdraw) {
+                // Withdraw tokens from dydx to the `otherAccount`.
+                actions[i] = _createWithdrawAction(
+                    withdrawTo,
+                    scaledAmount,
+                    bridgeAction
+                );
+            } else {
+                // If all values in the `Action` enum are handled then this
+                // revert is unreachable: Solidity will revert when casting
+                // from `uint8` to `Action`.
+                revert("DydxBridge/UNRECOGNIZED_BRIDGE_ACTION");
+            }
+        }
+    }
+
+    /// @dev Returns a dydx `DepositAction`.
+    /// @param depositFrom Deposit tokens from this address who is also the account owner.
+    /// @param amount of tokens to deposit.
+    /// @param bridgeAction A `BridgeAction` struct.
+    /// @return depositAction The encoded dydx action.
+    function _createDepositAction(
+        address depositFrom,
+        uint256 amount,
+        BridgeAction memory bridgeAction
+    )
+        internal
+        pure
+        returns (
+            IDydx.ActionArgs memory depositAction
+        )
+    {
+        // Create dydx amount.
+        IDydx.AssetAmount memory dydxAmount = IDydx.AssetAmount({
+            sign: true,                                 // true if positive.
+            denomination: IDydx.AssetDenomination.Wei,  // Wei => actual token amount held in account.
+            ref: IDydx.AssetReference.Target,           // Target => an absolute amount.
+            value: amount                               // amount to deposit.
+        });
+
+        // Create dydx deposit action.
+        depositAction = IDydx.ActionArgs({
+            actionType: IDydx.ActionType.Deposit,           // deposit tokens.
+            amount: dydxAmount,                             // amount to deposit.
+            accountId: bridgeAction.accountId,              // index in the `accounts` when calling `operate`.
+            primaryMarketId: bridgeAction.marketId,         // indicates which token to deposit.
+            otherAddress: depositFrom,                      // deposit from the account owner.
```

Comment:
> Does dydx have any kind of whitelisting for which addresses can deposit into accounts?
Yes, see [here](https://github.com/dydxprotocol/solo/blob/master/contracts/protocol/impl/OperationImpl.sol#L316)
(i) The `otherAddress` must be either the sender or account owner, and
(ii) The sender must be an account operator.

> Is it practical for mesh, etc, to validate that the depositFrom address is somehow linked to the accountId?

The `accountId` here is actually just an offset into the `accounts` array (see [here](https://github.com/dydxprotocol/solo/blob/master/contracts/protocol/lib/Actions.sol#L284)), which we construct in this contract. A client-side check could validate that `from` is the account owner and that this bridge is a valid operator. However, we don't perform these checks here because dydx will revert in this scenario.

Reply:
Ah, great :+1: